### PR TITLE
Add Latch data source

### PR DESCRIPTION
**Pull request by AidanAbd**
# TL;DR
_Please replace this text with a description of what this PR accomplishes._

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/lyft/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/lyft/flyte/issues/<number>_

### CODE REVIEW

File: flytekit/interfaces/data/latch/latch_proxy.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        data = r.json()
+        presigned_urls = data["urls"]
+        upload_id = data["upload_id"]
+        f = open(file_path, "rb")
+        parts=[]
+        for key, val in presigned_urls.items():
+            blob = f.read(self._chunk_size)
+            r = requests.put(val, data=blob)
+            if r.status_code != 200:
+                raise _FlyteUserException("failed to upload part `{}` of file `{}`".format(key, file_path))
```
[the remark this answers is not in the document]
Comment (AidanAbd) [reply flytekit/interfaces/data/latch/latch_proxy.py:123]: Already does in the outer data api

File: flytekit/interfaces/data/latch/latch_proxy.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    def raw_output_data_prefix_override(self) -> str:
+        return self._raw_output_data_prefix_override
+
+    @staticmethod
+    def _split_s3_path_to_bucket_and_key(path):
+        """
+        :param Text path:
+        :rtype: (Text, Text)
+        """
+        path = path[len("latch://") :]
```
[the remark this answers is not in the document]
Comment (AidanAbd) [reply flytekit/interfaces/data/latch/latch_proxy.py:42]: done

File: flytekit/interfaces/data/latch/latch_proxy.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        
+        return r.json()["exists"]
+
+    def download_directory(self, remote_path, local_path):
+        """
+        :param Text remote_path: remote latch:/// path
+        :param Text local_path: directory to copy to
+        """
+        if not remote_path.startswith("latch:///"):
+            raise ValueError("not a Latch URL. Please use URL of the format latch:///...")
```
Comment (maximsmol) [inline flytekit/interfaces/data/latch/latch_proxy.py:67]:
```suggestion
            raise ValueError(f"expected a Latch URL (latch:///...): {remote_path}")
```

Comment (maximsmol) [reply flytekit/interfaces/data/latch/latch_proxy.py:67]: likewise elsewhere

File: flytekit/interfaces/data/latch/latch_proxy.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            parts.append({'ETag': etag, 'PartNumber': int(key) + 1})
+        
+        r = requests.post(self._latch_endpoint + "/api/complete-upload", json={"upload_id": upload_id, "parts": parts, "object_url": to_path, "execution_name": _os.environ.get("FLYTE_INTERNAL_EXECUTION_ID")})
+        if r.status_code != 200:
+            raise _FlyteUserException("failed to complete upload for `{}`".format(to_path))
+        return True
+
+    def upload_directory(self, local_path, remote_path):
+        """
+        :param Text local_path:
```
Comment (maximsmol) [inline flytekit/interfaces/data/latch/latch_proxy.py:135]:
```suggestion
        :param str local_path:
```

likewise elsewhere

**Review by maximsmol — CHANGES_REQUESTED** [review]

File: flytekit/interfaces/data/latch/latch_proxy.py (hunk 4 above)
Comment (AidanAbd) [reply flytekit/interfaces/data/latch/latch_proxy.py]: done

**Pull request merged**